File: Stiver_Problems/Graphs/P017.cpp
// P017.cpp - Count enclaved '1' cells

#include <bits/stdc++.h>
using namespace std;

// Brute: DFS from each '1' to see if region touches border; count internal cells
int brute(const vector<vector<char>>& grid) {
    int row = grid.size();
    int col = grid[0].size();
    vector<vector<bool>> visited(row, vector<bool>(col, false));

    function<bool(int,int,int&)> dfs = [&](int r, int c, int &cnt) -> bool {
        if (r == 0 || c == 0 || r == row-1 || c == col-1) return true;
        visited[r][c] = true;
        cnt++;
        static int dr[4] = {0,1,-1,0};
        static int dc[4] = {1,0,0,-1};
        bool touch = false;
        for (int k = 0; k < 4; k++) {
            int nr = r + dr[k], nc = c + dc[k];
            if (0 <= nr && nr < row && 0 <= nc && nc < col &&
                !visited[nr][nc] && grid[nr][nc] == '1') {
                if (dfs(nr, nc, cnt)) touch = true;
            }
        }
        return touch;
    };

    int total = 0;
    for (int i = 0; i < row; i++) {
        for (int j = 0; j < col; j++) {
            if (grid[i][j] == '1' && !visited[i][j]) {
                int cnt = 0;
                bool touches = dfs(i, j, cnt);
                if (!touches) total += cnt;
            }
        }
    }
    return total;
}

// Optimal: BFS from all border '1's, remaining '1's are enclaves
int optimal(vector<vector<char>> grid) {
    int row = grid.size();
    int col = grid[0].size();
    vector<vector<bool>> visited(row, vector<bool>(col, false));
    queue<pair<int,int>> q;

    auto push_if1 = [&](int r, int c) {
        if (grid[r][c] == '1' && !visited[r][c]) {
            visited[r][c] = true;
            q.push({r, c});
        }
    };

    for (int j = 0; j < col; j++) {
        push_if1(0, j);
        push_if1(row - 1, j);
    }
    for (int i = 1; i < row - 1; i++) {
        push_if1(i, 0);
        push_if1(i, col - 1);
    }

    static int dr[4] = {1,0,-1,0};
    static int dc[4] = {0,1,0,-1};
    while (!q.empty()) {
        auto [r, c] = q.front(); q.pop();
        for (int k = 0; k < 4; k++) {
            int nr = r + dr[k], nc = c + dc[k];
            if (0 <= nr && nr < row && 0 <= nc && nc < col &&
                grid[nr][nc] == '1' && !visited[nr][nc]) {
                visited[nr][nc] = true;
                q.push({nr, nc});
            }
        }
    }

    int count = 0;
    for (int i = 0; i < row; i++)
        for (int j = 0; j < col; j++)
            if (grid[i][j] == '1' && !visited[i][j]) count++;
    return count;
}

int main() {
    vector<vector<char>> grid = {
        {'0','0','0','0'},
        {'1','0','1','0'},
        {'0','1','1','0'},
        {'0','0','0','0'}
    };

    cout << "Brute: " << brute(grid) << "\n";
    cout << "Optimal: " << optimal(grid) << "\n";
    return 0;
}

File: Graphs/P017.py
def brute(grid):
    row,col=len(grid),len(grid[0])
    visited=[[False]*col for _ in range(row)]
    count=[0]
    def dfs(r,c):
        if r==0 or c==0 or r==row-1 or c==col-1:
            return True
        visited[r][c]=True
        count[0]+=1
        for dr,dc in [(0,1),(1,0),(-1,0),(0,-1)]:
            nr,nc=r+dr,c+dc
            if not visited[nr][nc] and grid[nr][nc]=='1':
                return dfs(nr,nc)
        return False
    tot=0
    for i in range(row):
        for j in range(col):
            if grid[i][j]=='1' and not visited[i][j]:
                count[0]=0
                if not dfs(i,j):
                    tot+=count[0]
    return tot

from collections import deque
def optimal(grid):
    row,col=len(grid),len(grid[0])
    q=deque()
    visited=[[False]*col for _ in range(row)]
    for i in range(col):
        if grid[0][i]=='1':
            q.append((0,i))
            visited[0][i]=True
        if grid[row-1][i]=='1':
            q.append((row-1,i))
            visited[row-1][i]=True
    for i in range(1,row-1):
        if grid[i][0]=='1':
            q.append((i,0))
            visited[i][0]=True
        if grid[i][col-1]=='1':
            q.append((i,col-1))
            visited[i][col-1]=True
    while q:
        r,c,=q.popleft()
        for dr,dc in [(0,1),(1,0),(-1,0),(0,-1)]:
            nr,nc=r+dr,c+dc
            if 0<=nr<row and 0<=nc<col and grid[nr][nc]=='1' and not visited[nr][nc]:
                visited[nr][nc]=True
                q.append((nr,nc))
    count=0
    for i in range(row):
        for j in range(col):
            if grid[i][j]=='1' and not visited[i][j]:
                count+=1
    return count

if __name__=="__main__":
    grid=[['0','0','0','0'],
          ['1','0','1','0'],
          ['0','1','1','0'],
          ['0','0','0','0']]
    print("Brute: ",brute(grid))
    print("Optimal: ",optimal(grid))